File: tests/pipewire_impulse.py
"""Exercise all channels on a separate PipeWire core, without headset/hardware access."""
import array
import json
import os
from pathlib import Path
import signal
import math
import random
import struct
import subprocess as sp
import sys
import tempfile
import time
sys.path.insert(0,str(Path(__file__).resolve().parents[1]/'src'))
from build_graph import build
from sony_filters import CHANNELS
ROOT=Path(__file__).resolve().parents[1]

def main():
    processes=[]
    with tempfile.TemporaryDirectory(prefix='inzone-dsp-test-') as tmp:
        tmp=Path(tmp)
        env=dict(os.environ,PIPEWIRE_RUNTIME_DIR=str(tmp),XDG_RUNTIME_DIR=str(tmp),PIPEWIRE_REMOTE='inzone-test')
        graph=build(ROOT/'assets')
        graph['playback.props'].pop('target.object')
        graph['playback.props']['node.passive']=False
        graph['playback.props']['node.always-process']=True
        config={
          'context.properties':{'core.daemon':True,'core.name':'inzone-test','default.clock.rate':48000,'default.clock.quantum':256},
          'context.spa-libs':{'audio.convert.*':'audioconvert/libspa-audioconvert','support.*':'support/libspa-support'},
          'context.modules':[{'name':'libpipewire-module-'+m} for m in ('protocol-native','access','metadata','spa-node-factory','client-node','adapter','link-factory')]+[{'name':'libpipewire-module-filter-chain','args':graph}],
          'context.objects':[{'factory':'spa-node-factory','args':{'factory.name':'support.node.driver','node.name':'Dummy-Driver','node.group':'pipewire.dummy','priority.driver':200000}}]}
        def spa(value):
            if isinstance(value,dict):return '{ '+ ' '.join(json.dumps(k)+' = '+spa(v) for k,v in value.items())+' }'
            if isinstance(value,list):return '[ '+' '.join(spa(v) for v in value)+' ]'
            return json.dumps(value,ensure_ascii=False)
        (tmp/'server.conf').write_text('\n'.join(k+' = '+spa(v) for k,v in config.items()))
        log=(ROOT/'analysis/pipewire-impulse.log').open('w')
        def start(args,**kwargs):
            p=sp.Popen(args,env=env,stderr=log,**kwargs);processes.append(p);return p
        def run(*args):return sp.check_output(args,env=env,text=True,stderr=log,timeout=8)
        def wait_for(predicate):
            for _ in range(100):
                if predicate():return
                time.sleep(.05)
            (ROOT/'analysis/test-dump.json').write_text(run('pw-dump'))
            raise RuntimeError('Timed out waiting for isolated graph')
        try:
            daemon=start(['pipewire','-c',str(tmp/'server.conf')])
            wait_for(lambda:(tmp/'inzone-test').exists() or daemon.poll() is not None)
            if daemon.poll() is not None:raise RuntimeError('Test server failed to start')
            # Eight distinct impulses, one channel every 4096 samples, after one second silence.
            frames=48000+8*4096+96000
            samples=array.array('f',[0])*(frames*8)
            for index in range(8):samples[(48000+index*4096)*8+index]=0.125
            rng=random.Random(2987)
            start_signal=48000+8*4096
            for i in range(48000):
                level=(.05,.4,2.,.1)[i//12000]
                for ch in range(8):samples[(start_signal+i)*8+ch]=level*rng.uniform(-1,1)
            (tmp/'input.raw').write_bytes(samples.tobytes())
            record=(tmp/'output.raw').open('wb')
            rec=start(['pw-cat','-r','--raw','--format','f32','--rate','48000','--channels','2','--target','0','--latency','256','-P','{ node.name = test-record node.always-process = true }','-'],stdout=record)
            player=start(['pw-cat','-p','--raw','--format','f32','--rate','48000','--channels','8','--channel-map',','.join(CHANNELS),'--target','0','--latency','256','-P','{ node.name = test-play }',str(tmp/'input.raw')])
            wait_for(lambda:any(n.get('info',{}).get('props',{}).get('node.name')=='test-play' for n in json.loads(run('pw-dump'))))
            for n in json.loads(run('pw-dump')):
                if n['type']!='PipeWire:Interface:Node':continue
                name=n['info']['props'].get('node.name')
                if name not in ('test-play','test-record','inzone.sony-surround','inzone.sony-surround.output'):continue
                direction='Output' if name in ('test-play','inzone.sony-surround.output') else 'Input'
                positions=list(CHANNELS) if name in ('test-play','inzone.sony-surround') else ['FL','FR']
                param={'direction':direction,'mode':'dsp','format':{'mediaType':'audio','mediaSubtype':'raw','format':'F32P','rate':48000,'channels':len(positions),'position':positions}}
                run('pw-cli','set-param',str(n['id']),'PortConfig',json.dumps(param))
            wait_for(lambda:'test-record' in run('pw-link','-i') and 'test-play' in run('pw-link','-o'))
            for ch in ('FL','FR'):run('pw-link',f'inzone.sony-surround.output:output_{ch}',f'test-record:input_{ch}')
            for ch in CHANNELS:run('pw-link',f'test-play:output_{ch}',f'inzone.sony-surround:playback_{ch}')
            player.wait(timeout=12)
            time.sleep(.2)
            rec.send_signal(signal.SIGINT);rec.wait(timeout=3);record.close()
            values=array.array('f');values.frombytes((tmp/'output.raw').read_bytes())
            if not values or max(map(abs,values))<1e-5:raise AssertionError('No output from graph')
            # Independently check eight audible DSP responses, finite limiting,
            # and recovery into silence. No external implementation is used.
            assert all(math.isfinite(v) for v in values)
            first=next(i for i,v in enumerate(values[::2]) if abs(v)>1e-5)
            peaks=[]
            for index in range(8):
                start=max(0,first+index*4096-128);end=first+index*4096+2048
                segment=values[start*2:end*2]
                peaks.append(max(map(abs,segment)))
            assert all(v>1e-5 for v in peaks),peaks
            peak=max(map(abs,values));assert peak<=1.000001,peak
            tail=max(map(abs,values[-24000:]));assert tail<1e-5,tail
            result={'samples':len(values),'channel_impulse_peaks':peaks,'peak':peak,'silence_tail_peak':tail,'checks':'finite output, eight channel responses, ALC bound and silence recovery'}
            (ROOT/'analysis/impulse-results.json').write_text(json.dumps(result,indent=2)+'\n')
            print('PASS: Linux PipeWire spatial processing:',result)
        finally:
            for p in reversed(processes):
                if p.poll() is None:
                    p.terminate()
                    try:p.wait(timeout=3)
                    except sp.TimeoutExpired:p.kill();p.wait()
            log.close()

if __name__=='__main__':main()

File: Sources/InzoneDiagnostics/PipeWireImpulse.swift
import Foundation
import Glibc
import InzoneCore

public enum PipeWireDiagnostics {
    public static func runImpulse(repository: URL) throws -> [String: Any] {
        let analysis = repository.appendingPathComponent("analysis")
        let session = try IsolatedPipeWireSession(prefix: "inzone-dsp-test", remote: "inzone-test",
                                                  logURL: analysis.appendingPathComponent("pipewire-impulse.log"))
        defer { session.close() }
        var graph = try GraphRenderer(paths: InzonePaths()).buildSurround(
            assets: repository.appendingPathComponent("assets"), plugin: repository.appendingPathComponent("native/inzone_dsp.so"))
        guard var playback = graph["playback.props"] as? [String: Any] else { throw InzoneError.message("Surround playback properties are missing.") }
        playback.removeValue(forKey: "target.object")
        playback["node.passive"] = false
        playback["node.always-process"] = true
        graph["playback.props"] = playback
        var server = session.baseServer()
        var modules = server["context.modules"] as? [[String: Any]] ?? []
        modules.append(["name": "libpipewire-module-filter-chain", "args": graph])
        server["context.modules"] = modules
        try session.writeServer(server)
        let daemon = try session.start(["pipewire", "-c", session.serverURL.path])
        try session.waitFor(description: "isolated PipeWire server") {
            FileManager.default.fileExists(atPath: session.socketURL.path) || !daemon.isRunning
        }
        guard daemon.isRunning else { throw InzoneError.message("The isolated PipeWire server failed to start.") }

        let frames = 48000 + 8 * 4096 + 96000
        var samples = [Float](repeating: 0, count: frames * 8)
        for channel in 0..<8 { samples[(48000 + channel * 4096) * 8 + channel] = 0.125 }
        var generator = DiagnosticRandom(seed: 2987)
        let signalStart = 48000 + 8 * 4096
        let levels: [Double] = [0.05, 0.4, 2, 0.1]
        for frame in 0..<48000 {
            for channel in 0..<8 {
                samples[(signalStart + frame) * 8 + channel] = Float(levels[frame / 12000] * generator.uniform())
            }
        }
        let inputURL = session.directory.appendingPathComponent("input.raw")
        let outputURL = session.directory.appendingPathComponent("output.raw")
        try FloatSamples.encode(samples).write(to: inputURL)
        let output = try session.outputFile(outputURL)
        defer { try? output.close() }
        let recorder = try session.start(["pw-cat", "-r", "--raw", "--format", "f32", "--rate", "48000",
            "--channels", "2", "--target", "0", "--latency", "256", "-P",
            "{ node.name = test-record node.always-process = true }", "-"], output: output)
        let player = try session.start(["pw-cat", "-p", "--raw", "--format", "f32", "--rate", "48000",
            "--channels", "8", "--channel-map", GraphRenderer.channels.joined(separator: ","),
            "--target", "0", "--latency", "256", "-P", "{ node.name = test-play }", inputURL.path])
        try session.waitFor(description: "isolated playback and recording nodes") {
            let nodes = try session.nodes()
            return nodes["test-play"] != nil && nodes["test-record"] != nil
        }
        for (name, node) in try session.nodes() {
            guard ["test-play", "test-record", "inzone.sony-surround", "inzone.sony-surround.output"].contains(name),
                  let identifier = node["id"] as? NSNumber else { continue }
            let outputDirection = name == "test-play" || name == "inzone.sony-surround.output"
            let positions = name == "test-play" || name == "inzone.sony-surround" ? GraphRenderer.channels : ["FL", "FR"]
            let parameter = Self.portConfiguration(output: outputDirection, positions: positions)
            _ = try session.run(["pw-cli", "set-param", identifier.stringValue, "PortConfig", JSONSupport.encode(parameter, pretty: false)])
        }
        try session.waitFor(description: "isolated DSP ports") {
            let inputs = try session.run(["pw-link", "-i"])
            let outputs = try session.run(["pw-link", "-o"])
            return inputs.contains("test-record") && outputs.contains("test-play")
        }
        for channel in ["FL", "FR"] {
            _ = try session.run(["pw-link", "inzone.sony-surround.output:output_\(channel)", "test-record:input_\(channel)"])
        }
        for channel in GraphRenderer.channels {
            _ = try session.run(["pw-link", "test-play:output_\(channel)", "inzone.sony-surround:playback_\(channel)"])
        }
        try session.wait(player, timeout: 12)
        Thread.sleep(forTimeInterval: 0.2)
        if recorder.isRunning { _ = Glibc.kill(recorder.processIdentifier, SIGINT) }
        try session.wait(recorder, timeout: 3, requireSuccess: false)
        try output.synchronize()
        try output.seek(toOffset: 0)
        let values = try FloatSamples.decode(try output.readToEnd() ?? Data())
        guard !values.isEmpty, values.allSatisfy(\.isFinite),
              let first = stride(from: 0, to: values.count, by: 2).first(where: { abs(values[$0]) > 0.00001 }).map({ $0 / 2 }) else {
            throw InzoneError.message("The isolated surround graph produced no finite audible output.")
        }
        var peaks: [Double] = []
        for channel in 0..<8 {
            let start = max(0, first + channel * 4096 - 128) * 2
            let end = min(values.count, (first + channel * 4096 + 2048) * 2)
            guard start < end else { throw InzoneError.message("The surround recording is missing channel \(channel).") }
            peaks.append(Double(values[start..<end].map(abs).max() ?? 0))
        }
        let peak = Double(values.map(abs).max() ?? 0)
        let tail = Double(values.suffix(24000).map(abs).max() ?? 0)
        guard peaks.allSatisfy({ $0 > 0.00001 }), peak <= 1.000001, tail < 0.00001 else {
            throw InzoneError.message("Surround response, ALC bound, or silence recovery validation failed: peaks=\(peaks), peak=\(peak), tail=\(tail).")
        }
        let result: [String: Any] = ["samples": values.count, "channel_impulse_peaks": peaks, "peak": peak,
            "plugin_sha256": try Digests.sha256(file: repository.appendingPathComponent("native/inzone_dsp.so")),
            "silence_tail_peak": tail, "checks": "finite output, eight channel responses, ALC bound and silence recovery"]
        try Self.writeJSON(result, to: analysis.appendingPathComponent("impulse-results.json"))
        return result
    }

    static func portConfiguration(output: Bool, positions: [String]) -> [String: Any] {
        ["direction": output ? "Output" : "Input", "mode": "dsp", "format": [
            "mediaType": "audio", "mediaSubtype": "raw", "format": "F32P", "rate": 48000,
            "channels": positions.count, "position": positions,
        ]]
    }

    static func writeJSON(_ value: Any, to destination: URL) throws {
        try AtomicFile.write(Data((JSONSupport.encode(value) + "\n").utf8), to: destination)
    }
}

final class IsolatedPipeWireSession {
    let directory: URL
    let remote: String
    var serverURL: URL { directory.appendingPathComponent("server.conf") }
    var socketURL: URL { directory.appendingPathComponent(remote) }
    private var environment: [String: String]
    private let log: FileHandle
    private var cancellation: IsolatedDiagnosticCancellation?
    private var processes: [Process] = []
    private var closed = false

    init(prefix: String, remote: String, logURL: URL) throws {
        directory = FileManager.default.temporaryDirectory.appendingPathComponent("\(prefix)-\(UUID().uuidString)")
        self.remote = remote
        try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: false, attributes: [.posixPermissions: 0o700])
        do {
            try FileManager.default.createDirectory(at: logURL.deletingLastPathComponent(), withIntermediateDirectories: true)
            let stagedLog = try AtomicFile.stage(for: logURL)
            log = try stagedLog.publishAndTakeFileHandle()
        } catch {
            try? FileManager.default.removeItem(at: directory)
            throw error
        }
        var values = ProcessInfo.processInfo.environment
        values["PIPEWIRE_RUNTIME_DIR"] = directory.path
        values["XDG_RUNTIME_DIR"] = directory.path
        values["PIPEWIRE_REMOTE"] = remote
        values["XDG_CONFIG_HOME"] = directory.appendingPathComponent("config").path
        // Clients still require the installed client.conf while the daemon receives an explicit private configuration.
        for key in ["PIPEWIRE_CONFIG_DIR", "PIPEWIRE_CONFIG_NAME", "PIPEWIRE_CONFIG_PREFIX"] { values.removeValue(forKey: key) }
        environment = values
        cancellation = IsolatedDiagnosticCancellation()
    }

    deinit { close() }

    func stagePlugin(_ source: URL) throws -> String {
        let name = "inzone_dsp_" + (try Digests.sha256(file: source)).prefix(16)
        let libraryDirectory = directory.appendingPathComponent("ladspa")
        try FileManager.default.createDirectory(at: libraryDirectory, withIntermediateDirectories: true)
        let destination = libraryDirectory.appendingPathComponent(name + ".so")
        try FileManager.default.copyItem(at: source, to: destination)
        // The unique basename and private search path prevent an installed plugin from substituting for the build under test.
        environment["LADSPA_PATH"] = libraryDirectory.path
        return name
    }

    func close() {
        guard !closed else { return }
        closed = true
        for process in processes.reversed() { stop(process) }
        processes.removeAll()
        try? log.close()
        try? FileManager.default.removeItem(at: directory)
        cancellation = nil
    }

    func baseServer() -> [String: Any] {
        ["context.properties": ["core.daemon": true, "core.name": remote, "default.clock.rate": 48000, "default.clock.quantum": 256],
         "context.spa-libs": ["audio.convert.*": "audioconvert/libspa-audioconvert", "support.*": "support/libspa-support"],
         "context.modules": ["protocol-native", "access", "metadata", "spa-node-factory", "client-node", "adapter", "link-factory"].map { ["name": "libpipewire-module-" + $0] },
         "context.objects": [["factory": "spa-node-factory", "args": ["factory.name": "support.node.driver", "node.name": "Dummy-Driver", "node.group": "pipewire.dummy", "priority.driver": 200000]]]]
    }

    func writeServer(_ config: [String: Any]) throws {
        let text = try config.keys.sorted().map { key in "\(key) = \(try Self.spa(config[key]!))" }.joined(separator: "\n")
        try Data(text.utf8).write(to: serverURL)
    }

    private static func spa(_ value: Any) throws -> String {
        if let object = value as? [String: Any] {
            return try "{ " + object.keys.sorted().map { key in
                try JSONSupport.encode(key, pretty: false) + " = " + spa(object[key]!)
            }.joined(separator: " ") + " }"
        }
        if let array = value as? [Any] { return try "[ " + array.map(spa).joined(separator: " ") + " ]" }
        return try JSONSupport.encode(value, pretty: false)
    }

    func outputFile(_ url: URL) throws -> FileHandle {
        let stagedOutput = try AtomicFile.stage(for: url)
        return try stagedOutput.publishAndTakeFileHandle()
    }

    func start(_ arguments: [String], output: FileHandle? = nil) throws -> Process {
        try checkCancellation()
        let process = Process()
        process.executableURL = URL(fileURLWithPath: "/usr/bin/env")
        process.arguments = arguments
        process.environment = environment
        process.standardInput = FileHandle.nullDevice
        process.standardOutput = output ?? FileHandle.nullDevice
        process.standardError = log
        try process.run()
        processes.append(process)
        return process
    }

    func run(_ arguments: [String], timeout: TimeInterval = 8) throws -> String {
        let outputURL = directory.appendingPathComponent("command-\(UUID().uuidString)")
        let output = try outputFile(outputURL)
        defer { try? output.close(); try? FileManager.default.removeItem(at: outputURL) }
        let process = try start(arguments, output: output)
        try wait(process, timeout: timeout)
        try output.seek(toOffset: 0)
        return String(decoding: try output.readToEnd() ?? Data(), as: UTF8.self)
    }

    func wait(_ process: Process, timeout: TimeInterval, requireSuccess: Bool = true) throws {
        let deadline = ProcessInfo.processInfo.systemUptime + timeout
        while process.isRunning && ProcessInfo.processInfo.systemUptime < deadline {
            try checkCancellation()
            Thread.sleep(forTimeInterval: 0.01)
        }
        try checkCancellation()
        guard !process.isRunning else {
            stop(process)
            throw InzoneError.message("Diagnostic command timed out: \(process.arguments?.joined(separator: " ") ?? "process")")
        }
        process.waitUntilExit()
        guard !requireSuccess || process.terminationStatus == 0 else {
            throw InzoneError.message("Diagnostic command exited with status \(process.terminationStatus): \(process.arguments?.joined(separator: " ") ?? "process")")
        }
    }

    func waitFor(timeout: TimeInterval = 5, description: String, condition: () throws -> Bool) throws {
        let deadline = ProcessInfo.processInfo.systemUptime + timeout
        while ProcessInfo.processInfo.systemUptime < deadline {
            try checkCancellation()
            if try condition() { return }
            Thread.sleep(forTimeInterval: 0.05)
        }
        throw InzoneError.message("Timed out waiting for \(description).")
    }

    func checkCancellation() throws {
        guard isolatedDiagnosticCancelled == 0 else { throw InzoneError.message("Private PipeWire diagnostics were cancelled.") }
    }

    func nodes() throws -> [String: [String: Any]] {
        guard let snapshot = try JSONSupport.decode(Data(run(["pw-dump"]).utf8)) as? [[String: Any]] else {
            throw InzoneError.message("Invalid private PipeWire graph snapshot.")
        }
        var result: [String: [String: Any]] = [:]
        for node in snapshot where node["type"] as? String == "PipeWire:Interface:Node" {
            if let info = node["info"] as? [String: Any], let properties = info["props"] as? [String: Any],
               let name = properties["node.name"] as? String { result[name] = node }
        }
        return result
    }

    private func stop(_ process: Process) {
        guard process.isRunning else { return }
        process.terminate()
        let deadline = ProcessInfo.processInfo.systemUptime + 3
        while process.isRunning && ProcessInfo.processInfo.systemUptime < deadline { Thread.sleep(forTimeInterval: 0.01) }
        if process.isRunning { _ = Glibc.kill(process.processIdentifier, SIGKILL) }
        process.waitUntilExit()
    }
}

nonisolated(unsafe) private var isolatedDiagnosticCancelled: sig_atomic_t = 0
private func isolatedDiagnosticSignalHandler(_ signal: Int32) { isolatedDiagnosticCancelled = 1 }

private final class IsolatedDiagnosticCancellation {
    private let previousInterrupt: sig_t?
    private let previousTermination: sig_t?

    init() {
        isolatedDiagnosticCancelled = 0
        previousInterrupt = Glibc.signal(SIGINT, isolatedDiagnosticSignalHandler)
        previousTermination = Glibc.signal(SIGTERM, isolatedDiagnosticSignalHandler)
    }

    deinit {
        _ = Glibc.signal(SIGINT, previousInterrupt)
        _ = Glibc.signal(SIGTERM, previousTermination)
    }
}
